ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,265 @@
+```text
+ACP: 99
+Title: ValidatorSetManager Solidity Contract
+Author(s): Gauthier Leonard <https://github.com/Nuttymoon>
+Discussions-To: https://github.com/avalanche-foundation/ACPs/discussions/98
+Status: Proposed
+Track: Best Practices
+```
+
+## Abstract
+
+Define (i) a reference implementation for a minimal Solidity smart contract to manage a Subnet’s validator set from an Avalanche EVM chain, as well as (ii) a reference architecture to easily plug in custom “security modules” on top of this contract (e.g. to implement a PoS Subnet).
+
+The proposed name for this contract is `ValidatorSetManager`.
+
+This proposal should also help define the payload of the P-Chain notifications sent to the Subnet upon validator addition and removal.
+
+This ACP relies on concepts introduced in [ACP-73 (Warp Addressed Transactions)](https://github.com/avalanche-foundation/ACPs/pull/73) and [ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets). It depends on them to transition to an `Implementable` state.
+
+## Motivation
+
+[ACP-77 (Reinventing Subnets)](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets) opens the door to managing a Subnet validator set (stored on the P-Chain) from another chain of the Avalanche network. This is achieved by sending updates to the P-Chain from a “validator manager” address on this chain. The “validator manager” address will be the only one able to issue `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` on the P-Chain to add, change the weight of, and remove validators.
+
+`Warp Addressed Transactions` (introduced in [ACP-73](https://github.com/avalanche-foundation/ACPs/pull/73)) could be used to send the validator set updates to the P-Chain.
+
+Upon addition or removal of a validator, the P-Chain is expected to send a Warp message back to the Subnet. The Subnet needs a way to interpret this message and trigger appropriate actions. These notification messages have not been specified to date.
+
+Given these assumptions and the fact that most of the active blockchains on Avalanche mainnet are EVM-based (especially the C-Chain, on which we can expect most of the Subnet “validator manager” addresses will be hosted for security reasons), we propose defining a reference implementation for a Solidity smart contract that can:
+
+1. Hold relevant information about the current Subnet validator set, as well as historical information
```

Comment:
If we place ourselves in the context of a PoS xL1, when a validator leaves the validator set we should be able to compute staking rewards based on the historical evolution of its weight and uptime. It is easier to do if all this information is directly queryable on-chain, rather than having to do an asynchronous query to the P-Chain. This is the whole purpose of the `SubnetValidationPeriod` struct defined below.

That being said, tracking the total validator set weight of the xL1 over time would be pretty cumbersome, since it changes at every single validator weight change.